sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/aggregateFunctions.scala

```diff
@@ -2023,3 +2023,136 @@ case class GpuVarianceSamp(child: Expression, nullOnDivideByZero: Boolean)
 
   override def prettyName: String = "var_samp"
 }
+
+case class GpuReplaceNullmask(
+    input: Expression, 
+    mask: Expression) extends GpuExpression with ShimExpression {
+
+  override def dataType: DataType = input.dataType
+  override def nullable: Boolean = mask.nullable
+  override def children: Seq[Expression] = Seq(input, mask)
+
+  override def columnarEval(batch: ColumnarBatch): GpuColumnVector = {
+    val maskColumnNotNull = withResource(mask.columnarEval(batch)) { maskColumn =>
+      maskColumn.getBase.isNotNull
+    }
+    val res = withResource(maskColumnNotNull) { _ =>
+      withResource(input.columnarEval(batch)) { inputColumn =>
+        inputColumn.getBase.copyWithBooleanColumnAsValidity(maskColumnNotNull)
+      }
+    }
+    GpuColumnVector.from(res, dataType)
+  }
+}
+
+object CudfMaxMinBy {
+  val KEY_ORDERING: String = "_key_ordering"
+  val KEY_VALUE: String = "_key_value"
+}
+
+abstract class CudfMaxMinByAggregate(
+    orderingType: DataType,
+    valueType: DataType) extends CudfAggregate {
+
+  protected val reductionAggregation: ReductionAggregation
+
+  override lazy val reductionAggregate: cudf.ColumnVector => cudf.Scalar = col => {
+    if (col.getNullCount == col.getRowCount) { // all nulls
+      GpuScalar.from(null, dataType)
+    } else {
+      withResource(col.copyToColumnVector()) { colVec =>
```

Comment:
AFAIK, the input "col" is already a ColumnVector, why does it still need to call "copyToColumnVector" here ?

Reply:
Thanks, fixed, it doesn't.